tsl/src/compression/api.c

```diff
@@ -892,3 +976,649 @@ tsl_recompress_chunk_wrapper(Chunk *uncompressed_chunk)
+}
+
+static bool
+recompress_remote_chunk(FunctionCallInfo fcinfo, Chunk *chunk)
+{
+	return invoke_compression_func_remotely(fcinfo, chunk);
+}
+
+/*
+ * This is hacky but it doesn't matter. We just want to check for the existence of such an index
+ * on the compressed chunk. For distributed hypertables, returning the index oid would raise issues,
+ * because the Access Node does not see that oid. So we return the oid of the uncompresed chunk
+ * instead, when an index is found.
+ */
+extern Datum
+tsl_get_compressed_chunk_index_for_recompression(PG_FUNCTION_ARGS)
+{
+	Oid uncompressed_chunk_id = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	Chunk *uncompressed_chunk = ts_chunk_get_by_relid(uncompressed_chunk_id, true);
+
+	if (NULL == uncompressed_chunk)
+		elog(ERROR, "unknown chunk id %d", uncompressed_chunk_id);
+
+	/* push down to data nodes for distributed case */
+	if (uncompressed_chunk->relkind == RELKIND_FOREIGN_TABLE)
+	{
+		/* look for index on data nodes */
+		if (!(invoke_compression_func_remotely(fcinfo, uncompressed_chunk)))
+			PG_RETURN_NULL();
+		else // don't care what the idx oid is, data node will find it and open it
+			PG_RETURN_OID(uncompressed_chunk_id);
+	}
+	int32 srcht_id = uncompressed_chunk->fd.hypertable_id;
+	Chunk *compressed_chunk = ts_chunk_get_by_id(uncompressed_chunk->fd.compressed_chunk_id, true);
+
+	List *htcols_list = ts_hypertable_compression_get(srcht_id);
+	ListCell *lc;
+	int htcols_listlen = list_length(htcols_list);
+
+	const ColumnCompressionInfo **colinfo_array;
+	colinfo_array = palloc(sizeof(ColumnCompressionInfo *) * htcols_listlen);
+
+	int nsegmentby_cols = 0, i = 0;
+
+	foreach (lc, htcols_list)
+	{
+		FormData_hypertable_compression *fd = (FormData_hypertable_compression *) lfirst(lc);
+		colinfo_array[i++] = fd;
+		if (COMPRESSIONCOL_IS_SEGMENT_BY(fd))
+			nsegmentby_cols++;
+	}
+
+	const ColumnCompressionInfo **keys;
+	int n_keys;
+	int16 *in_column_offsets = compress_chunk_populate_keys(uncompressed_chunk->table_id,
+															colinfo_array,
+															htcols_listlen,
+															&n_keys,
+															&keys);
+
+	Relation compressed_chunk_rel = table_open(compressed_chunk->table_id, AccessShareLock);
+	Relation uncompressed_chunk_rel = table_open(uncompressed_chunk->table_id, AccessShareLock);
+	TupleDesc compressed_rel_tupdesc = RelationGetDescr(compressed_chunk_rel);
+	TupleDesc uncompressed_rel_tupdesc = RelationGetDescr(uncompressed_chunk_rel);
+
+	RowCompressor row_compressor;
+	row_compressor_init(&row_compressor,
+						uncompressed_rel_tupdesc,
+						compressed_chunk_rel,
+						htcols_listlen,
+						colinfo_array,
+						in_column_offsets,
+						compressed_rel_tupdesc->natts,
+						true /*need_bistate*/,
+						true /*recompressing*/);
+
+	if (OidIsValid(row_compressor.index_oid))
+	{
+		/* lock in least strict mode so it's not dropped by other sessions */
+		Relation index_rel = relation_open(row_compressor.index_oid, AccessExclusiveLock);
```

Comment:
Thats not least strict mode

Reply:
Wouldnt we want that lock on the chunk and not the index, why are we locking the index specifically?

Reply:
Yes the comment is misleading, it was there before I changed the AccessShareLock to AccessExclusive and I forgot to adjust it. I'll fix that. 
The AccessExclusiveLock on the index is taken because eventually, we will reindex the compressed chunk relation, which will request an AccessExclusiveLock. Requesting it in the beginning of the transaction prevents a deadlock between two concurrent recompress operations. 

Then again maybe a deadlock there is not so unacceptable as one of the recompression operations should fail anyway.